src/garage/_environment.py

```diff
@@ -0,0 +1,226 @@
+    | spec                  | The environment specifications                  |
+    +-----------------------+-------------------------------------------------+
+    | render_modes          | The list of supported render modes              |
+    +-----------------------+-------------------------------------------------+
+
+    Example of a simple rollout loop:
+
+    .. code-block:: python
+
+        env = MyEnv()
+        policy = MyPolicy()
+        first_ts = env.reset()
+        env.visualize()  # visualization window opened
+
+        episode = [first_ts, ]
+        while not ts.last():
+            action = policy.get_action(episode[-1].next_observation)
+            episode.append(env.step(action))
+
+        env.close()  # visualization window closed
+
+    Make sure your environment is pickle-able:
+        Garage pickles the environment via the `cloudpickle` module
+        to save snapshots of the experiment. However, some environments may
+        contain attributes that are not pickle-able (e.g. a client-server
+        connection). In such cases, override `__setstate__()` and
+        `__getstate()__` to add your custom pickle logic.
+
+        You might want to refer to the EzPickle module:
+        https://github.com/openai/gym/blob/master/gym/utils/ezpickle.py
+        for a lightweight way of pickle and unpickle via constructor
+        arguments.
+
+    """
+
+    @property
+    @abc.abstractmethod
+    def action_space(self):
+        """Return the action space specification.
+
+        Some environments do not provide their action space specifications as
+        type `akro.Space`. In such cases, subclasses should silently convert
+        the environment's action_space to `akro.Space`.
+
+        Returns:
+            akro.Space: The environment specification.
+
+        """
+
+    @property
+    @abc.abstractmethod
+    def observation_space(self):
+        """Return the observation space specification.
+
+        Some environments do not provide their observation space
+        specifications as type `akro.Space`. In such cases, subclasses should
+        silently convert the environment's observation_space to `akro.Space`.
+
+        Returns:
+            akro.Space: The environment specification.
+        """
+
+    @property
+    @abc.abstractmethod
+    def spec(self):
+        """Return the environment specification.
+
+        Returns:
+            garage.EnvSpec: The environment specification.
+
+        """
+
+    @property
+    @abc.abstractmethod
+    def render_modes(self):
+        """Return the list of supported render modes.
+
+        Returns:
+            list: A list of string representing the supported render modes.
+            See render() for a list of modes.
+
+        """
+
+    @abc.abstractmethod
+    def reset(self):
+        """Resets the environment and returns the first `TimeStep` of sequence.
+
+        Returns:
+            garage.TimeStep: The first time step. Note that this `TimeStep`
```

Comment:
generally in the case of garage types you can omit the `garage.foo.bar` stem. For external code, include the full package path.